wandb/sdk/launch/runner/kubernetes_runner.py

```diff
@@ -64,11 +66,132 @@
 }
 
 
+def _is_preempted(status: "V1PodStatus") -> bool:
+    """Check if this pod has been preempted."""
+    if hasattr(status, "conditions") and status.conditions is not None:
+        for condition in status.conditions:
+            if condition.type == "DisruptionTarget" and condition.reason in [
+                "EvictionByEvictionAPI",
+                "PreemptionByScheduler",
+                "TerminationByKubelet",
+            ]:
+                return True
+    return False
+
+
+class KubernetesRunMonitor:
+    def __init__(
+        self,
+        job_field_selector: str,
+        pod_label_selector: str,
+        namespace: str,
+        batch_api: "BatchV1Api",
+        core_api: "CoreV1Api",
+    ) -> None:
+        """Initial KubernetesRunMonitor.
+
+        Arguments:
+            jobname: Name of the job.
+
+        Returns:
+            None.
+        """
+        self.pod_label_selector = pod_label_selector
+        self.job_field_selector = job_field_selector
+        self.namespace = namespace
+        self.batch_api = batch_api
+        self.core_api = core_api
+
+        self._status_lock = Lock()
+        self._status = Status("starting")
+
+        self._watch_job_thread = Thread(target=self._watch_job, daemon=True)
+        self._watch_pods_thread = Thread(target=self._watch_pods, daemon=True)
+
+    def start(self) -> None:
+        """Start the run monitor."""
+        if self._watch_job_thread.is_alive() or self._watch_pods_thread.is_alive():
+            raise LaunchError(
+                "Attempted to start monitor that has already started"
+            )  # TODO: what should I do here?
+        self._watch_job_thread.start()
+        self._watch_pods_thread.start()
+
+    def _set_status(self, status: Status) -> None:
+        """Set the run status."""
+        with self._status_lock:
+            self._status = status
+
+    def get_status(self) -> Status:
+        """Get the run status."""
+        with self._status_lock:
+            return self._status
+
+    def _watch_pods(self) -> None:
+        """Watch for pods created matching the jobname."""
+        watcher = watch.Watch()
+        try:
+            # Start stream with a small timeout to check if the pods have been created
+            # within the time limit
+            for event in watcher.stream(
+                self.core_api.list_namespaced_pod,
+                namespace=self.namespace,
+                label_selector=self.pod_label_selector,
+                _request_timeout=10,  # TODO: make this a configurable constant
+            ):
+                # We just want to know that the pods have been created within time limit
+                if event.get("type") in ["ADDED", "MODIFIED"]:
+                    break
+
```

Comment:
@TimH98 this is where the timeout for pod creation is implemented

Reply:
So by the time a pod is in ContainerCreating it's already out of this loop?

Reply:
Yeah as soon as the pod resource is created this will exit. Which is also what `wait_job_launch` was testing for I believe